src/features/device-state/routes/state-patch-v3.ts

```diff
@@ -0,0 +1,384 @@
+import type { Filter } from 'pinejs-client-core';
+import { metricsPatchFields, v3ValidPatchFields } from '..';
+import { shouldUpdateMetrics } from '../utils';
+
+const { BadRequestError, UnauthorizedError, InternalRequestError } = errors;
+const { api } = sbvrUtils;
+
+/**
+ * These typings should be used as a guide to what should be sent, but cannot be trusted as what actually *is* sent.
+ */
+export type StatePatchV3Body = {
+	[uuid: string]: {
+		name?: string;
+		status?: string;
+		os_version?: string;
+		os_variant?: string;
+		supervisor_version?: string;
+		provisioning_progress?: number | null;
+		provisioning_state?: string | null;
+		ip_address?: string;
+		mac_address?: string;
+		api_port?: number;
+		api_secret?: string;
+		logs_channel?: string; // TODO: should this not be reported anymore?
+		memory_usage?: number;
+		memory_total?: number;
+		storage_block_device?: string;
+		storage_usage?: number;
+		storage_total?: number;
+		cpu_temp?: number;
+		cpu_usage?: number;
+		cpu_id?: string;
+		is_undervolted?: boolean;
+		/**
+		 * Used for setting dependent devices as online
+		 */
+		is_online?: boolean;
+		/**
+		 * Used for setting gateway device of dependent devices
+		 */
+		is_managed_by__device?: number;
+		apps?: {
+			[uuid: string]: {
+				/**
+				 * We report the overall release uuid the supervisor considers the active one, even though there may be info for multiple releases.
+				 */
+				release_uuid?: string;
+				releases?: {
+					[releaseUUID: string]: {
+						services?: {
+							[name: string]: {
+								image: string;
+								status: string;
+								download_progress?: number;
+							};
+						};
+					};
+				};
+			};
+		};
+	};
+};
+
+export const statePatchV3: RequestHandler = async (req, res) => {
+	try {
+		const body = req.body as StatePatchV3Body;
+		const custom: AnyObject = {}; // shove custom values here to make them available to the hooks
+
+		// forward the public ip address if the request is from the supervisor.
+		if (req.apiKey != null) {
+			custom.ipAddress = getIP(req);
+		}
+
+		const uuids = Object.keys(body).filter((uuid) => body[uuid] != null);
+		if (uuids.length === 0) {
+			throw new BadRequestError();
+		}
+
+		const appReleaseUuids: {
+			[appUuid: string]: Set<string>;
+		} = {};
+		const imageLocations: string[] = [];
+		for (const uuid of uuids) {
+			const { apps } = body[uuid];
+			if (apps != null) {
+				for (const [
+					appUuid,
+					{ release_uuid: isRunningReleaseUuid, releases },
+				] of Object.entries(apps)) {
+					appReleaseUuids[appUuid] ??= new Set();
+					if (isRunningReleaseUuid) {
+						appReleaseUuids[appUuid].add(isRunningReleaseUuid);
+					}
+					if (releases != null) {
+						for (const [releaseUuid, { services }] of Object.entries(
+							releases,
+						)) {
+							appReleaseUuids[appUuid].add(releaseUuid);
+							if (services != null) {
+								for (const [, service] of Object.entries(services)) {
```

Comment:
I guess it could be:
```suggestion
								for (const service of Object.values(services)) {
```